### DISCUSSION TITLE
Does SpinMutex from spin crate not work on aarch64?

### DISCUSSION THREAD

**Discussion by RedDocMD** [rust-embedded/rust-raspberrypi-OS-tutorials#132] [Tutorial Questions] [answered]
Hi!
Instead of using the `NullLock` defined in this tutorial, I am using the spin-lock from the `spin` crate. However, it seems that the lock is never acquired and it always blocks forever. Is there some specific setup required to use `spin` on Raspberry Pi 3A+?

**Comment by jakubfijalkowski** [comment, marked as answer]
Spin lock uses atomics under the hood, but atomics work on ARM only if you have MMU configured and enabled. MMU is introduced later than the `NullLock`, so I assume that is the case (it was in my case).

The forever spinning isn't really spinning - CPU raises an exception, but you must have not configured the handler yet so it raises an exception from an exception handler, freezing forever (cause handlers are introduced even after MMU).

**Comment by andre-richter** [comment]
What @jakubfijalkowski says. On the RPi you get an exception, if you have them enabled already, if you try atomics on non-cacheable memory.

There is also a hint starting tutorial 10 in `main.rs`:

https://github.com/rust-embedded/rust-raspberrypi-OS-tutorials/blob/ea8ef5a7f1ae31837cf49a3a1ee691c907d0a7df/10_virtual_mem_part1_identity_mapping/src/main.rs#L134-L142

Implementation of exclusive load/store (atomics) is often tied to / implemented on top of the CPU’s cache (leveraging the HW-coherency it provides).
